### PR TITLE
feat: add user hash for intercom

### PR DESCRIPTION
**Pull request by G3root**
Make Intercom more secure

Created via Threads. See full discussion: https://threads.com/34464425134

### CODE REVIEW

**Comment by hariombalhara** [comment]
@G3root What exactly is it trying to fix? need a PR description

File: apps/web/pages/api/intercom-hash.ts [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+    case "GET":
+      if (id.success) {
+        // eslint-disable-next-line turbo/no-undeclared-env-vars
+        const hmac = crypto.createHmac("sha256", process.env.INTERCOM_SECRET ?? "");
+        hmac.update(id.data);
+        const hash = hmac.digest("hex");
+
+        res.status(200).json({ hash });
+      } else {
+        res.status(404).json({ message: "user id is required" });
```
Comment (hariombalhara) [inline apps/web/pages/api/intercom-hash.ts:20]:
```suggestion
        res.status(400).json({ message: "user id is required" });
```

File: apps/web/pages/api/intercom-hash.ts [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+export default function handler(req: NextApiRequest, res: NextApiResponse) {
+  const { query, method } = req;
+
+  const id = z.string().safeParse(query?.id);
+
+  switch (method) {
+    case "GET":
+      if (id.success) {
+        // eslint-disable-next-line turbo/no-undeclared-env-vars
+        const hmac = crypto.createHmac("sha256", process.env.INTERCOM_SECRET ?? "");
```
Comment (hariombalhara) [inline apps/web/pages/api/intercom-hash.ts:14]: It looks like a new env variable. It should be added in .env.example with comment on when/why it should be set.

File: apps/web/pages/api/intercom-hash.ts [hunk 3]
```diff
@@ -0,0 +1,28 @@
+import type { NextApiRequest, NextApiResponse } from "next";
+import crypto from "node:crypto";
+import { z } from "zod";
+
+export default function handler(req: NextApiRequest, res: NextApiResponse) {
+  const { query, method } = req;
+
+  const id = z.string().safeParse(query?.id);
+
+  switch (method) {
```
Comment (hariombalhara) [inline apps/web/pages/api/intercom-hash.ts:10]: We should use `defaultHandler` that automatically handles this switch case

File: packages/features/ee/support/lib/intercom/useIntercom.ts [hunk 4]
```diff
@@ -25,12 +25,23 @@ export const useIntercom = () => {
   const { hasPaidPlan } = useHasPaidPlan();
   const { hasTeamPlan } = useHasTeamPlan();
 
-  const open = () => {
+  const open = async () => {
+    let userHash;
+
+    if (data && data.id) {
```
Comment (hariombalhara) [inline packages/features/ee/support/lib/intercom/useIntercom.ts:31]:
```suggestion
    if (data?.id) {
```

File: apps/web/pages/api/intercom-hash.ts [hunk 5]
```diff
@@ -0,0 +1,28 @@
+import type { NextApiRequest, NextApiResponse } from "next";
+import crypto from "node:crypto";
+import { z } from "zod";
+
+export default function handler(req: NextApiRequest, res: NextApiResponse) {
+  const { query, method } = req;
```
Comment (hariombalhara) [inline apps/web/pages/api/intercom-hash.ts:6]: This endpoint seems open and anyone can send any user ID. Shouldn't we restrict it to userId of the session? In that case, `userId` can be derived from session itself.

**Review by hariombalhara — CHANGES_REQUESTED** [review]
Left some comments

**Review by hariombalhara — APPROVED** [review]

File: turbo.json [hunk 6]
```diff
@@ -277,6 +277,7 @@
     "ZOHOCRM_CLIENT_ID",
     "ZOHOCRM_CLIENT_SECRET",
     "ZOOM_CLIENT_ID",
-    "ZOOM_CLIENT_SECRET"
+    "ZOOM_CLIENT_SECRET",
+    "INTERCOM_SECRET"
```
Comment (zomars) [inline turbo.json:281]: Can we sort this alphabetically?

Comment (PeerRich) [reply turbo.json:281]: ah shit, didnt see this. maybe in a follow up PR?

Comment (zomars) [reply turbo.json:281]: Nah, it's ok

**Pull request merged**